File: freee/mergesort.py
def merge_sort(a):
    if len(a) <= 1:
        return a
    
    mid = len(a) // 2
    # print(mid)
    left = merge_sort(a[:mid])
    right = merge_sort(a[mid:])
    return merge(left,right)

#coding: utf-8
def merge(left,right):
    # n = len(left) + len(right)

    i = 0
    j = 0
    A = []
    # for k in range(0,n):
    #     if left[i] <= right[j]:
    #         A.append(left[i])
    #         i = i+1
    #     else:
    #         A.append(right[j])
    #         j = j+1
    while i < len(left) and j < len(right):
        # print(left[i],"and i", i)
        # print(right[j],"and j", j, right)
        if left[i] <= right[j]:
            A.append(left[i])
            i = i + 1
        else:
            A.append(right[j])
            j = j + 1

    A.extend(left[i:])
    A.extend(right[j:])

    return A 


if __name__ == '__main__':
    Arr = [2,4,1,65,22,6,7,34,5,8]
    sorted_arr = merge_sort(Arr)
    print(sorted_arr)

File: freee/MergeSort.java
import java.util.Arrays;

// import sun.awt.www.content.audio.aiff;
// import sun.security.util.Length;

public class MergeSort{
    public static int[] mergeSort(int[] a, int left, int right){
        int mid = (left + right) / 2;
        if (left < right) {
            mergeSort(a, left, mid);
            mergeSort(a, mid+1, right);
            
            // merge(a, left, mid, right);
            merge(a, mid, left, right);
            
            // System.out.println(Arrays.toString(a));
        }
        // System.out.println(Arrays.toString(a));
        return a;
    }

    public static void merge(int[] a, int mid,int left, int right){
        int i = left; int j = mid + 1;int k = 0;
        // int mid = (left + right) / 2;
        int lenth = right - left +1;
        // System.out.println("right "+right);
        int res[] = new int[lenth];
        
        while(i <= mid && j <= right){
            // System.out.println(right);
            // System.out.println(i+" i and j "+j);
            if (a[i] < a[j] ){
                res[k++] = a[i++];
            } else {
                res[k++] = a[j++];
            }
        }

        while(i <= mid){
            // System.out.println("res" + res[k]);
            // System.out.println("a" + a[i]);
            res[k++] = a[i++];
        }
        
        while(j <= right){
            // System.out.println("res2" +" "+ k);
            // System.out.println("a2" +" "+ j);
            res[k++] = a[j++];
        }
        
        for(int x=0; x<res.length;x++){
            a[x+left] = res[x];
        }
    }

    public static void main(String[] args){
        int a[] = {2,4,1,65,22,6,7,34,8,5,13,552,44,66};
        // int[] res = mergeSort(a, 0, a.length-1); 
        int[] res = mergeSort(a, 0, a.length-1);

        System.out.println(Arrays.toString(res));
    }
}